#  Bias-Variance & Diagnosing Model Fit

<hr style="height: 5px; background-color: #5A98C4 ; border: none;">

### Table of Contents:

1. [Introduction](#intro)
2. [Section 1 — Bias, Variance & Diagnosing Model Fit](#section1)
3. [Section 2 — Diagnosing Bias and Variance](#section2)
4. [Section 3 — Bias–Variance Trade-off ](#section3)
5. [Section 4 — Regularization](#section4)
6. [Section 5 — Hands-on Lap](#section5)
7. [References](#refs)

<hr style="height: 5px; background-color: #5A98C4 ; border: none;">

<a id="intro"></a>
# Introduction

In [Day 1 ](https://github.com/thimarArda/BinX_AI_-_ML_Training/blob/main/Week4_Evaluating_and_Tuning/Day1_Train_Validation_Test_Split/Train_Validation_Test_Splits.ipynb) and [Day 2 ](https://github.com/thimarArda/BinX_AI_-_ML_Training/blob/main/Week4_Evaluating_and_Tuning/Day2_Cross_Validation/Day2_Cross_Validation.ipynb) notebooks, we focused on getting a reliable performance estimate and choosing the best model using the validation set. We also discussed how not having a good distribution in the validation set might lead to **overfitting** and give us a misleading idea about how well our model is actually performing.

Today, we're going to touch upon some very important concepts that will help us understand what is happening inside our model when it performs well or poorly. We will understand why the model is getting it's final score. Is the model too simple and unable to capture the patterns in the data? Or is it too complex and learning the training data too well?

These questions lead us to two important concepts: **bias** and **variance**. Understanding them will help us recognize **underfitting** and **overfitting**, understand the trade-off between them, and know what kind of changes we should make to our model.

Finally, we will introduce the concept of regularization, By taking a look at two regularization methods,  L1 and L2 regularization, and how they can help us control the complexity of a model to reduce overfitting.

<hr style="height: 5px; background-color: #5A98C4 ; border: none;">

<a id="section1"></a>
# Section 1 : Bias, Variance & Diagnosing Model Fit

> This section explains bias and variance, two key concepts used to diagnose whether a machine learning model is underfitting or overfitting, and to guide improvements in model performance.


### It is important to remember that,
Datasets are divided into **training, development (dev), and test sets** in machine learning. The training set is used to train the model, the dev set is used to tune and improve the model, and the test set is used to evaluate final performance on unseen data. 

## 1.1 What is Bias ?

**Bias** is the error that occurs when a model is too simple to capture the true patterns in the data.

- **High bias (underfitting):** The model oversimplifies, misses patterns and underfits the data.
- **Low bias:** The model captures patterns well and is closer to the true values.

This usually corresponds to **high bias and relatively low variance**.

### Underfitting
> When the decision boundary is too simple and cannot properly separate the two classes. As a result, many data points are classified incorrectly. This situation is called **underfitting** because the model fails to learn the underlying patterns in the data.
>   <div style="text-align: center;"> <img src="underfitting.png" alt="Illustration of Underfitting " width="300" > </div>

For example, imagine trying to predict house prices using a model that only considers one simple relationship when the real problem depends on many nonlinear relationships.

The model may perform poorly on both the training and validation data.

### Typical Pattern

| Dataset | Performance |
|---|---|
| Training data | Poor |
| Validation/Test data | Poor |

The important clue is that the model is **already performing poorly on the training data**.


  


---------
### How to Reduce Bias?
Some methods to lower bias in models are:

1. **Use More Complex Models:** Use models capable of capturing non-linear relationships such as neural networks(this is an advanced topic for later) or ensemble methods.
2. **Add Relevant Features:** Include additional informative features in the training data to give the model for capturing underlying patterns.
3. **Adjust Regularization Strength:** Reduce regularization to allow the model more flexibility in fitting the data.


----------

## 1.2 What is Variance ? 

**Variance** refers to how much a model's predictions change when it is trained on different training datasets. It performes very well on training data but fails at the unseen ones.


### Overfitting
> The boundary is very irregular and tries to fit almost every training example perfectly, including noise and unusual points. This causes **overfitting**, where the model performs well on training data but poorly on new unseen examples.
>  <div style="text-align: center;"> <img src="overfitting.png" alt="Illustration of Overfiting " width="300" > </div>

For example, a very deep decision tree may memorize the training examples instead of learning general rules.

### Typical Pattern

| Dataset | Performance |
|---|---|
| Training data | Very good |
| Validation/Test data | Poor |

**The large difference between training and validation performance is an important sign of overfitting.**

### How to Reduce Variance?
Some methods to lower variance are:

1. **Simplify the Model:** Use a simpler model or prune overly deep decision trees to avoid overfitting.
2. **Increase Training Data:** Collect more data to stabilize learning and make the model generalize better.
3. **Apply Regularization:** Use L1 or L2 regularization to constrain model complexity and prevent overfitting.
4. **Use Ensemble Methods:** Implement techniques like bagging or random forests to combine multiple models and balance bias–variance trade-offs.
5. **Use cross-validation**


<hr style="height: 5px; background-color: #5A98C4 ; border: none;">

<a id="section2"></a>
# Section 2 : Diagnosing Bias and Variance

## 2.1 Diagnosing with the model's performance

| | Underfitting (High Bias) | Overfitting (High Variance) |
|---|---|---|
| **Symptom** | Poor on both training **and** validation data | Great on training, poor on validation data |
| **Cause** | Model too simple to capture the real pattern | Model too complex; it memorized the training data instead of learning the general trend |
| **Fix** | Add features, use a more powerful/flexible model | Simplify the model, get more data, or **regularize** |

> **The two fixes are opposites.** Applying the overfitting fix (simplifying) to an underfit model makes it
> worse, and vice-versa — this is why diagnosis has to come *before* the fix.

------------

## Diagnosing With the Train-vs-Validation Gap

> **The gap between the training score and the validation score is the practical diagnostic tool** 

| Training Score | Validation Score | Diagnosis |
|---|---|---|
| Low | Low | **Underfitting** — model too simple |
| High | Much lower | **Overfitting** — model memorized the training data |
| High | High (small gap) | **Good fit** — this is the target |

**In words:** if the model can't even do well on the data it was trained on, it's too simple
(underfitting). If it does great on training data but falls apart on validation data, it has memorized
noise specific to the training set instead of the general pattern (overfitting). If both scores are high
*and* close together, that's the sweet spot from Section 1.2.



<hr style="height: 5px; background-color: #5A98C4 ; border: none;">

<a id="section3"></a>
# Section 3 : Bias–Variance Trade-off 

Bias and variance represent two different sources of model error.

In general:

- **Increasing model complexity** tends to decrease bias but increase variance.
- **Decreasing model complexity** tends to increase bias but decrease variance.

Therefore, we want to find a balance between them.

This is known as the **Bias–Variance Trade-off**.

Formally, for a model $\hat{f}(x)$ trying to estimate the true function $f(x)$ given noisy targets
$y = f(x) + \epsilon$ (with $\mathbb{E}[\epsilon] = 0$, $\text{Var}(\epsilon) = \sigma^2$), the expected
squared prediction error at a point $x$ decomposes into exactly three terms:

$$
\mathbb{E}\Big[\big(y - \hat{f}(x)\big)^2\Big] \;=\; \underbrace{\Big(\text{Bias}\big[\hat f(x)\big]\Big)^2}_{\text{Bias}^2}
\;+\; \underbrace{\text{Var}\big[\hat f(x)\big]}_{\text{Variance}} \;+\; \underbrace{\sigma^2}_{\text{Irreducible Error}}
$$

where:

$$
\text{Bias}\big[\hat f(x)\big] = \mathbb{E}\big[\hat f(x)\big] - f(x)
\qquad\qquad
\text{Var}\big[\hat f(x)\big] = \mathbb{E}\Big[\big(\hat f(x) - \mathbb{E}[\hat f(x)]\big)^2\Big]
$$

**Reading the decomposition:**
- **Bias²** — how far off the *average* prediction (across many re-trainings on different samples) is from
  the truth. High for underfit models.
- **Variance** — how much the prediction *swings* depending on which training sample was used. High for
  overfit models.
- **Irreducible error ($\sigma^2$)** — noise inherent to the data itself; no model, however good, can
  remove this term.

This is one of the most important ideas when tuning a machine learning model. It tells us that the model's total error comes from **three parts**: the error caused by **bias**, the error caused by **variance**, and some **irreducible error** that we cannot completely remove.

As we increase the model's complexity, bias usually **decreases** because the model becomes better at capturing patterns. However, variance usually **increases** because the model becomes more sensitive to the training data. The goal is to find the right level of complexity where the total error is as low as possible.

----------
### The Three General Cases

| Model | Bias | Variance | Typical Problem |
|---|---|---|---|
| Too simple | High | Low | Underfitting |
| Appropriate complexity | Balanced | Balanced | Good generalization |
| Too complex | Low | High | Overfitting |

The ideal model lies somewhere between the two extremes.


<hr style="height: 5px; background-color: #5A98C4 ; border: none;">

<a id="section4"></a>
# Section 4 : Regularization

## 4.1 What Regularization Does, and Why It Helps

The regularized logistic regression cost function can be written as:

$$
\boxed{
J(w,b)
=
\frac{1}{m}
\sum_{i=1}^{m}
\mathcal{L}\left(f_{w,b}(x^{(i)}),y^{(i)}\right)
+
\frac{\lambda}{2m}
\sum_{j=1}^{n}w_j^2
}
$$

Where:

- $J(w,b)$ → the total cost that the model tries to minimize
- $m$ → number of training examples
- $n$ → number of features
- $x^{(i)}$ → the features of the $i$-th training example
- $y^{(i)}$ → the actual target of the $i$-th training example
- $f_{w,b}(x^{(i)})$ → the model's prediction for the $i$-th example
- $\mathcal{L}$ → the original logistic regression loss for one example
- $w_j$ → the weight associated with feature $j$
- $b$ → the bias/intercept
- $\lambda$ → the regularization strength

### Breaking Down the Equation

The equation has two main parts:


$$
\underbrace{
\frac{1}{m}
\sum_{i=1}^{m}
\mathcal{L}\left(f_{w,b}(x^{(i)}), y^{(i)}\right)
}{\text{Original Logistic Regression Cost}}
+
\underbrace{
\frac{\lambda}{2m}
\sum{j=1}^{n} w_j^2
}_{\text{L2 Regularization Penalty}}
$$

The **first term** measures how well the model's predictions match the actual labels.

The **second term** is the regularization penalty. It adds a cost when the model's weights become large.

So, instead of only asking:

> *"How well can I fit the training data?"*

the model is also encouraged to ask:

> *"Can I fit the data without using unnecessarily large weights?"*

This discourages the model from becoming too complex and can help reduce overfitting.

---

### The Role of $\lambda$

The parameter $\lambda$ controls **how strongly we penalize large weights**.

- **Small $\lambda$** → weak regularization → weights can become larger → model is more flexible
- **Large $\lambda$** → strong regularization → weights are pushed more strongly toward zero → model becomes simpler

So, choosing $\lambda$ is another way of controlling the model's complexity.

> **Important:** If $\lambda$ is made too large, the model can become too simple and start underfitting. Therefore, the goal is not to use the strongest possible regularization, but to find a value that gives good generalization.

----------

## 4.2  Ridge Regression (L2 Regularization)

### How it works mathematically

L2 regularization adds a penalty based on the **squared values of the model's weights** to the cost function.

The L2 penalty is:

$$
\boxed{
\sum_{j=1}^{n} w_j^2
}
$$

Therefore, the regularized cost function becomes:

$$
J(w,b)
=
\frac{1}{m}
\sum_{i=1}^{m}
\mathcal{L}\left(f_{w,b}(x^{(i)}),y^{(i)}\right)
+
\frac{\lambda}{2m}
\sum_{j=1}^{n} w_j^2
$$

Because large weights produce a larger penalty, the model is encouraged to keep its weights **small**. L2 regularization usually does not make weights exactly zero; instead, it gradually shrinks them toward zero.

This makes L2 useful when I believe that **most of my features contain useful information**, but I want to prevent the model from relying too heavily on certain features.

> **L2 → squared weights → smaller weights → less complex and more stable model**


------

### Scikit-learn implementation:

```python
from sklearn.linear_model import Ridge

ridge = Ridge(alpha=1.0)   # alpha: regularization strength (larger = simpler model)
ridge.fit(X_train, y_train)
ridge.predict(X_val)
```

**Syntax explained:**
- `Ridge(alpha=1.0)` — creates the estimator. `alpha` is the only hyperparameter that matters for
  regularization strength here; it directly corresponds to $\alpha$ in the equation above.
- `.fit(X_train, y_train)` — solves for the coefficients $\theta$ that minimize $J_{ridge}(\theta)$ on the
  training data. This is the step where regularization actually takes effect — the penalty term is baked
  into the optimization itself, not applied afterward.
- `.predict(X_val)` — produces predictions using the already-fit (shrunken) coefficients; no penalty is
  involved at prediction time, only at fitting time.
  
---------
### When to use Ridge

when most/all features are expected to be at least somewhat relevant, when
features are correlated with each other (Ridge handles multicollinearity gracefully by spreading weight
across correlated features rather than picking one arbitrarily), or simply as a safe first regularizer to
try before reaching for Lasso.

---------------

## 4.3 Lasso Regression (L1 Regularization)

### How it works mathematically
L1 regularization works in a similar way as L2, but instead of squaring the weights, it uses their **absolute values**.

The L1 penalty is:

$$
\boxed{
\sum_{j=1}^{n} |w_j|
}
$$

The regularized cost function becomes:

$$
J(w,b)
=
\frac{1}{m}
\sum_{i=1}^{m}
\mathcal{L}\left(f_{w,b}(x^{(i)}),y^{(i)}\right)
+
\frac{\lambda}{2m}
\sum_{j=1}^{n} |w_j|
$$

The important difference is that L1 regularization can push some weights **all the way to exactly zero**.

When a weight becomes zero, its corresponding feature no longer contributes to the model. Because of this, L1 can perform a form of **feature selection** and produce a sparse model.

This can be useful when I have many features and I believe that some of them may not be useful for making predictions.

> **L1 → absolute weights → some weights can become zero → feature selection**

----------

### Scikit-learn implementation:

```python
from sklearn.linear_model import Lasso

lasso = Lasso(alpha=0.1)   # alpha: regularization strength (larger = more coefficients pushed to 0)
lasso.fit(X_train, y_train)
lasso.predict(X_val)

# Which coefficients did Lasso zero out?
import numpy as np
print("Number of features:", len(lasso.coef_))
print("Number of coefficients set to exactly 0:", np.sum(lasso.coef_ == 0))
```

**Syntax explained:**
- `Lasso(alpha=0.1)` — same pattern as `Ridge`, but the class implements the L1 penalty instead of L2.
  Note Lasso typically needs a *smaller* `alpha` than Ridge for a comparable effect, since the L1 penalty
  grows differently than L2 — always tune it rather than reusing the same value across the two.
- `.fit(X_train, y_train)` — solves for $\theta$ minimizing $J_{lasso}(\theta)$; this is a harder
  optimization problem than Ridge's (no closed-form solution because $|\theta_j|$ isn't differentiable at
  0), so Scikit-learn uses coordinate descent under the hood.
- `lasso.coef_` — the fitted coefficient array; checking `== 0` reveals exactly which features Lasso
  decided were not worth keeping.

---------

### When to use Lasso :
when many features are suspected to be irrelevant or redundant, and I'd like the
model itself to tell me which ones matter — the automatic feature selection is a genuine practical benefit
on high-dimensional data. The trade-off: if several features are highly correlated, Lasso tends to
arbitrarily keep one and zero out the rest, which can be less stable than Ridge's "shrink everything a
little" approach.

## L1 vs. L2

The main difference can be seen directly in their penalty terms:

$$
\text{L2:} \qquad \sum_{j=1}^{n} w_j^2
$$

$$
\text{L1:} \qquad \sum_{j=1}^{n} |w_j|
$$

| | L1 Regularization | L2 Regularization |
|---|---|---|
| Penalty | $\sum |w_j|$ | $\sum w_j^2$ |
| Effect | Can make weights exactly zero | Shrinks weights toward zero |
| Feature selection | Yes | Generally no |
| Result | Sparse model | Smaller, smoother weights |
| Common name | Lasso | Ridge |

A simple way to remember them:

> **L1 can remove features. L2 keeps the features but reduces their influence.**

<hr style="height: 5px; background-color: #5A98C4 ; border: none;">

<a id="section5"></a>
# Section 5 : Hands-on Lap

### 1. Imorting the dataset, `make_classification` from scikit-learn

Initially, I used the `Breast Cancer` dataset from `scikit-learn` for this practical part. However, I noticed that almost all of the models were performing very well, even when I changed their complexity. It was difficult to clearly see the difference between underfitting, overfitting, and the effect of regularization.

I then tried the `Heart Disease` dataset from the previous notebook, but I faced another issue. The dataset contained a large number of duplicate rows, which could cause the training and validation sets to contain identical or nearly identical examples. This could make the validation performance look better than the model's actual ability to generalize to unseen data, so I decided that it was not the best choice for today's experiments.

I decided to use `make_classification` from `scikit-learn`. Which is a utility that allows me to generate an customize a binary classification dataset with a specific number of samples and features, while also controlling factors such as the number of informative features, redundant features, class separation, and noise.

**The dataset contains:**
- 10000 samples
- 20 features
- target values of 1 and 0

In [52]:
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

X, y = make_classification(
    n_samples=10000, n_features=20, n_informative=8, n_redundant=4,
    n_clusters_per_class=2, flip_y=0.15, random_state=42   # 15% label noise
)
X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

print("Training:", X_train.shape)
print("Validation:", X_val.shape)
print("Test:", X_test.shape)

Training: (8000, 20)
Validation: (2000, 20)
Test: (9769, 14)


### 2. Split the dataset into training/ validation / testing

In [42]:
from sklearn.model_selection import train_test_split 
from sklearn.metrics import accuracy_score



# 1) hold out 20% as the final test set
X_temp, X_test, y_temp, y_test = train_test_split(
 X, y, test_size=0.2, random_state=42)

# 2) split the rest into train (60%) and validation (20%)
X_train, X_val, y_train, y_val = train_test_split(
    X_temp,
    y_temp,
    test_size=0.2,
    random_state=42,
    stratify=y_temp
)


print("Training:", X_train.shape)
print("Validation:", X_val.shape)
print("Test:", X_test.shape)

Training: (31258, 14)
Validation: (7815, 14)
Test: (9769, 14)


## Step 1: Deliberately overfit a model and confirm the large train-validation gap.

#### Training a Decision Tree model but adding an unlimited depth to it.
> `max_depth=None` --> the tree has no artificial limit on how deep it can grow

In [65]:
from sklearn.tree import DecisionTreeClassifier


overfit_tree = DecisionTreeClassifier(max_depth=None, random_state=42)
overfit_tree.fit(X_train, y_train)

train_score = overfit_tree.score(X_train, y_train)
val_score = overfit_tree.score(X_val, y_val)

print("Training accurecy:", train_score)
print("Validation accurecy:", val_score)



Training accurecy: 1.0
Validation accurecy: 0.745


#### Interpretation

Noticing that the model scored **100%** accuracy on training data, but **74.5%** on the validation data. This is a huge gap between the training and validation accuracy, **which shows a clear sign of overfitting**. The model managed to memorize and capture certain patterns and noises, and it got accustomed a lot to the training data that it failed to generalize to the unseen validation data..

## Step 2: Deliberately underfit a model (e.g. an overly simple one) and confirm both scores are low

In [64]:
from sklearn.tree import DecisionTreeClassifier


underfit_tree = DecisionTreeClassifier(max_depth= 1, random_state=42)
underfit_tree.fit(X_train, y_train)

train_score = underfit_tree.score(X_train, y_train)
val_score = underfit_tree.score(X_val, y_val)

print("Training accurecy:", train_score)
print("Validation accurecy:", val_score)



Training accurecy: 0.618625
Validation accurecy: 0.608


#### Interpretation

`max_depth=1`  made the Decision Tree very simple. The model achieved 61.9% accuracy on the training data and 60.8% on the validation data.

Both scores are  low and very close to each other, which shows **high bias and underfitting**. The model is too simple to capture the important patterns in the data, so it performs poorly even on the training set. Since it also performs similarly poorly on the validation set, the model is not learning enough from the data to generalize well.

This shows the opposite behavior of the previous model.

## Step 3: Apply regularization (or reduce complexity) to the overfit model and show the gap shrink

In [70]:

fixed_tree = DecisionTreeClassifier(max_depth=9, random_state=42)
fixed_tree.fit(X_train, y_train)

train_score = fixed_tree.score(X_train, y_train)
val_score = fixed_tree.score(X_val, y_val)

print("Training accurecy:", train_score)
print("Validation accurecy:", val_score)


Training accurecy: 0.863
Validation accurecy: 0.782


#### Interpretation

After applying a reasonable `max_depth` to the Decision Tree, the gap between the training and validation accuracy became smaller. The training accuracy decreased compared to the overfitted model, while the validation accuracy increased.

The gap is now much more reasonable: the model is not overfitting as much as the overfitting model, but it is also not performing poorly on both sets like the underfitted model.

This shows that choosing more appropriate model settings can give us better results. Reducing the model complexity by tuning the `max_depth` hyperparameter helped reduce overfitting and slightly improved the validation performance.



## Step 4: Document each diagnosis and fix with the score evidence in Markdown.

| Model    |  `max_depth` | Training Accuracy | Validation Accuracy |   Gap | Diagnosis                   |
| -------- | -----------: | ----------------: | ------------------: | ----: | --------------------------- |
| Overfit  |       `None` |             1.000 |               0.740 | 0.260 | High variance / Overfitting |
| Underfit |          `1` |             0.619 |               0.608 | 0.011 | High bias / Underfitting    |
| Tuned    | *your value* |             0.863 |               0.782 | 0.081 | Better balance              |

> The results show the relationship between model complexity and generalization. The unrestricted tree achieved perfect training accuracy but performed much worse on the validation set, showing overfitting. When the tree was made too simple, both training and validation performance were low, showing underfitting. After choosing a more reasonable `max_depth`, the gap became smaller and validation accuracy improved, showing a better balance between bias and variance.

<hr style="height: 5px; background-color: #5A98C4 ; border: none;">

<a id="refs"></a>
# References :

- Read more about Regularization : https://www.geeksforgeeks.org/machine-learning/regularization-in-machine-learning/
- How to use make_classification : https://proclusacademy.com/blog/sklearn_make_classification/